ilqr.py

```diff
@@ -343,17 +366,328 @@ def _forward_pass(self, L_last):
+        # derivatives at every time step and compute the error of the interpolation
+        # for debugging
+        DEBUG = False
+
+        if(DEBUG and self.derivs_interpolation.keypoint_method == 'adaptiveJerk'):
+            jerkProfile = self.calc_jerk_profile(x)
+            plt.title("jerk profile")
+            plt.plot(jerkProfile[:,0])
+            plt.show()
+
+        # Calculate keypoints over the trajectory
+        keyPoints = []
+        if(self.derivs_interpolation.keypoint_method == 'setInterval'):
+            keyPoints = self.get_keypoints_set_interval()
+        elif(self.derivs_interpolation.keypoint_method == 'adaptiveJerk'):
+            keyPoints = self.get_keypoints_adaptive_jerk(x, u)
+        elif(self.derivs_interpolation.keypoint_method == 'iterativeError'):
+            keyPoints = self.get_keypoints_iterative_error(x, u)
+            self.deriv_calculated_at_index = [False] * self.N
+        else:
+            raise Exception('unknown interpolation method')
+
+        self.percentage_derivs = (len(keyPoints) / (self.N - 1)) * 100
+
+        if(DEBUG):
+            for t in range(self.N-1):
+                self.fx_baseline[:,:,t], self.fu_baseline[:,:,t] = self._calc_dynamics_partials(x[:,t], u[:,t])
+
+        # Calculate derivatives at keypoints
+        if self.derivs_interpolation.keypoint_method != 'iterativeError':
+            for t in range(len(keyPoints)):
+                self.fx[:,:,keyPoints[t]], self.fu[:,:,keyPoints[t]] = self._calc_dynamics_partials(x[:,keyPoints[t]], u[:,keyPoints[t]])
+
+        # Interpolate derivatives if required
+        if not (self.derivs_interpolation.keypoint_method == 'setInterval' and self.derivs_interpolation.minN == 1):
+            self.interpolate_derivatives(keyPoints)
+        
+        indexX = 0
+        indexY = 3
+        error_fx, error_fu = self.calc_error_of_interpolation()
+
+        if(DEBUG):  
+            print(f'error fx: {error_fx} error fu:  {error_fu}')
+            plt.title(f"error fx row {indexX} col {indexY}")
+            plt.plot(self.fx[indexX,indexY,:], label="interpolations")
+            plt.plot(self.fx_baseline[indexX,indexY,:], label="baseline")
+            plt.legend()
+            plt.show()    
+
+        # if(DEBUG):
+        #     error_fx, error_fu = self.calc_error_of_interpolation()
+
+        #     print(f'error fx: {error_fx} error fu:  {error_fu}')
+        #     for i in range(self.n):
+        #         for j in range(self.n):
+        #             plt.title(f"error fx row {i} col {j}")
+        #             plt.plot(self.fx[i,j,:], label="interpolations")
+        #             plt.plot(self.fx_baseline[i,j,:], label="baseline")
+        #             plt.legend()
+        #             plt.show()
+
+    def get_keypoints_set_interval(self):
+        """
+        Computes keypoints over the trajectory at set intervals as specified by the
+        interpolation method
+
+        Updates:
```

Comment:
minor, don't need this `Updates: N/A` thing if the method doesn't update anything. You might consider a `Returns:` documentation string, however.

Reply:
Same thing with all the other new methods